src/Microsoft.AspNetCore.Server.Kestrel/Http/SocketInput.cs

```diff
@@ -120,9 +120,9 @@ public void IncomingComplete(int count, Exception error)
                 {
                     _awaitableError = error;
                 }
-
-                Complete();
             }
+
+            Complete();
```

Comment:
Complete cannot be moved outside of the lock because it sets `_awaitableState` to `_awaitableIsCompleted`. Somewhat paradoxically setting this state outside the lock can cause the consuming thread to stall.

The above code inside the lock advances the tail which can change the result of `examined.IsEnd` (which must also be checked inside the lock). This may cause `ConsumingComplete` to incorrectly set the `_awaitableState` to `_awaitableIsNotCompleted` when there is really more data to consume which in turn causes the following `await SocketInput` to stall.

This is [exactly what happened](https://github.com/aspnet/KestrelHttpServer/pull/686/commits/84a68208d0b32d227a547daa32d2b25cad48ee92) the [first time](https://github.com/aspnet/KestrelHttpServer/pull/525/commits/afe944c053c6c127623f5996c19a99e9fdd1f4ad) we tried to remove the locks in SocketInput. I'm now very weary of reducing the locking in this class.

There are some things that could obviously be removed from the lock with a little refactoring like actual calls to `_threadPool.Run`, but does that actually have a measurable perf impact making it worth bifurcating `SocketInput.Complete`?